### ISSUE TITLE
JDK 11.0.13 - BCXDHPublicKey cannot be cast to class XECPublicKey

### ISSUE THREAD

**Issue by azorahai067** [bcgit/bc-java#1086]
JDK 11.0.13

Similar issue at https://github.com/bcgit/bc-java/issues/881 but that one is resolved in 11.0.10 which I confirmed.
Is this possibly fixed in any upcoming version?

BC 1.70 but this is coming from embedded Tomcat 8.5.34. Issue gets resolved when I upgraded to Tomcat 9.0.55 (have not tried any lower version as of yet).

```
java.lang.ClassCastException: class org.bouncycastle.jcajce.provider.asymmetric.edec.BCXDHPublicKey cannot be cast to class java.security.interfaces.XECPublicKey (org.bouncycastle.jcajce.provider.asymmetric.edec.BCXDHPublicKey is in unnamed module of loader org.springframework.boot.loader.LaunchedURLClassLoader @300ffa5d; java.security.interfaces.XECPublicKey is in module java.base of loader 'bootstrap')
	at java.base/sun.security.ssl.XDHKeyExchange$XDHEPossession.<init>(XDHKeyExchange.java:107) ~[na:na]
	at java.base/sun.security.ssl.NamedGroup$XDHScheme.createPossession(NamedGroup.java:602) ~[na:na]
	at java.base/sun.security.ssl.NamedGroup$NamedGroupSpec.createPossession(NamedGroup.java:511) ~[na:na]
	at java.base/sun.security.ssl.NamedGroup.createPossession(NamedGroup.java:419) ~[na:na]
	at java.base/sun.security.ssl.ECDHKeyExchange$ECDHEPossessionGenerator.createPossession(ECDHKeyExchange.java:259) ~[na:na]
	at java.base/sun.security.ssl.SSLKeyExchange$T12KeyAgreement.createPossession(SSLKeyExchange.java:352) ~[na:na]
	at java.base/sun.security.ssl.SSLKeyExchange.createPossessions(SSLKeyExchange.java:84) ~[na:na]
	at java.base/sun.security.ssl.ServerHello$T12ServerHelloProducer.chooseCipherSuite(ServerHello.java:432) ~[na:na]
	at java.base/sun.security.ssl.ServerHello$T12ServerHelloProducer.produce(ServerHello.java:295) ~[na:na]
	at java.base/sun.security.ssl.SSLHandshake.produce(SSLHandshake.java:436) ~[na:na]
	at java.base/sun.security.ssl.ClientHello$T12ClientHelloConsumer.consume(ClientHello.java:1102) ~[na:na]
	at java.base/sun.security.ssl.ClientHello$ClientHelloConsumer.onClientHello(ClientHello.java:854) ~[na:na]
	at java.base/sun.security.ssl.ClientHello$ClientHelloConsumer.consume(ClientHello.java:813) ~[na:na]
	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392) ~[na:na]
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443) ~[na:na]
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1074) ~[na:na]
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1061) ~[na:na]
	at java.base/java.security.AccessController.doPrivileged(Native Method) ~[na:na]
	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:1008) ~[na:na]
	at org.apache.tomcat.util.net.SecureNioChannel.tasks(SecureNioChannel.java:423) ~[tomcat-embed-core-8.5.34.jar!/:8.5.34]
	at org.apache.tomcat.util.net.SecureNioChannel.handshakeUnwrap(SecureNioChannel.java:483) ~[tomcat-embed-core-8.5.34.jar!/:8.5.34]
	at org.apache.tomcat.util.net.SecureNioChannel.handshake(SecureNioChannel.java:238) ~[tomcat-embed-core-8.5.34.jar!/:8.5.34]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1475) ~[tomcat-embed-core-8.5.34.jar!/:8.5.34]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) ~[tomcat-embed-core-8.5.34.jar!/:8.5.34]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) ~[tomcat-embed-core-8.5.34.jar!/:8.5.34]
	at java.base/java.lang.Thread.run(Thread.java:829) ~[na:na]
```

**Comment by dghgit** [comment]
Which provider jar is in use here?

**Comment by dghgit** [comment]
Actually re-reading this, I think it's going to be that the older version of Tomcat isn't handling the multi-release jar correctly.

**Comment by dghgit** [comment]
Closing as evidently a multi-class issue.

**Comment by regiseduardogregol** [comment]
any resolution? @azorahai067

**Comment by azorahai067** [comment]
> any resolution? @azorahai067

I ended up upgrading Tomcat to 9.0.55 (to resolve security vulnerabilities with Tomcat 8) which resolves the issue while using BC 1.70 (it's resolved starting in 1.69). If you are looking for resolution for JDK 11.0.13 and Tomcat 8.5.34, then I don't have an answer for you. I didn't find one in the short time I had to look into it.

**Comment by regiseduardogregol** [comment]
Thank you for sharing. My problem is with a Apache Karaf https://github.com/bcgit/bc-java/issues/1096
